### PR TITLE
Allow sloped/transform shape with non-identity transformation

### LINKED ISSUE #1058: `sloped` should consider the current transformation
**Brief outline of the proposed feature**

Inspired by [this SX question](https://tex.stackexchange.com/questions/618522/) and [this older one](tex.stackexchange.com/questions/184225/),
I would like to suggest that we should consider the current transformation
when computing the slope of the `sloped` nodes.

**Usage example**

The following code

```latex
\documentclass{article}
\usepackage{tikz}
\begin{document}
\begin{tikzpicture}[xscale=2, yscale=3, xshift=40, yshift=50]
	\draw (0, 0) -- node[sloped, above] {sloped correctly?} (1, 1);
\end{tikzpicture}
\end{document}
```

produces a line and a node that do not align.
This is because in the definition of `\pgftransformlineattime`,
the slope is calculated using the start and end coordinates
without applying the current transformation.
(For instance, in the above example, the diff vector will be (1cm, 1cm), not (2cm, 3cm).)
If we take that into account, a proper slope can be computed.

See [my answer](https://tex.stackexchange.com/a/618531/) for one possible fix.

### PR DESCRIPTION
**Pull request by Qrrbrbirlbel**
This applies the fixes of @thinbold and @muzimuzhi also to curves and arc while not globalize xc/yc for the line function.

In my opinion, this is not a breaking change because the previous behavior was never the intended way.

This is a redo of #1287 because I'm fighting with GitHub.

**Motivation for this change**

Fixes #1058

**Checklist**

Please [signoff your commits][git-s] to explicitly state your agreement to the [Developer Certificate of Origin][DCO]. If that is not possible you may check the boxes below instead:

- [X] Code changes are licensed under [GPLv2][GPL] + [LPPLv1.3c][LPPL]
- [x] Documentation changes are licensed under [FDLv1.2][FDL]

[git-s]: https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--s
[DCO]: https://developercertificate.org
[GPL]: https://www.gnu.org/licenses/gpl-2.0.html
[LPPL]: https://www.latex-project.org/lppl/lppl-1-3c.txt
[FDL]: https://www.gnu.org/licenses/fdl-1.2.html

### CODE REVIEW

**Comment by hmenke** [comment]
On 2026-06-15 a more comprehensive testsuite was introduced for PGF/TikZ. This tests all new contributions against the examples extracted from the manual and checks whether anything has changed to avoid regressions. Since this pull request was opened before that, these tests were not run automatically. I have run `l3build check` on the result of the merge of this pull request branch into `master`. The result can be found below.

> [!TIP]
> Your pull request might have failed some tests. This does not mean that your contribution will be rejected. In fact, depending on the changes you propose, it might even be expected that some tests fail, however, this needs to be evaluated carefully.

<details><summary>Failed tests</summary>

  - <details open><summary><code>pdftex</code></summary>

      - <details open><summary><code>config-examples</code></summary>

          - `pgfmanual-en-library-decorations`

        </details>

    </details>
  - <details open><summary><code>latexdvips</code></summary>

      - <details open><summary><code>config-examples</code></summary>

          - `pgfmanual-en-library-decorations`

        </details>

    </details>
  - <details open><summary><code>latexdvisvgm</code></summary>

      - <details open><summary><code>config-examples</code></summary>

          - `pgfmanual-en-library-decorations`

        </details>

    </details>
  - <details open><summary><code>luatex</code></summary>

      - <details open><summary><code>config-examples</code></summary>

          - `pgfmanual-en-library-decorations`

        </details>

    </details>
  - <details open><summary><code>xetex</code></summary>

      - <details open><summary><code>config-examples</code></summary>

          - `pgfmanual-en-library-decorations`

        </details>

    </details>

</details>

**Comment by Qrrbrbirlbel** [comment]
> On 2026-06-15 a more comprehensive testsuite was introduced for PGF/TikZ. This tests all new contributions against the examples extracted from the manual and checks whether anything has changed to avoid regressions. Since this pull request was opened before that, these tests were not run automatically. I have run `l3build check` on the result of the merge of this pull request branch into `master`. The result can be found below.

Is there an L3build for Dummies? How do I use this? I've run `l3build check` on my version of this branch and, yes, I've got even more failures …

I don't know to debug and inspect `pgfmanual-en-libraries-decorations` directly.

**Comment by hmenke** [comment]
Running the entire testsuite probably doesn't work for you because it requires a little bit of setup. It must be the latest TeX Live 2026 and gnuplot must be available.

To run only the failing test `pgfmanual-en-library-decorations` from the `examples` testsuite for the LuaTeX engine you can use
```console
$ l3build check -c config-examples -e luatex pgfmanual-en-library-decorations
[...lots of output...]

  Check failed with difference files
  - ./build/test-config-examples/pgfmanual-en-library-decorations.luatex.diff
```
The `diff` file contains more info about which tests have failed (`pgfmanual-en-library-decorations` contains 81 tests).

In case you have Podman (or Docker) installed, you can also reproduce the exact setup from the CI by running the tests inside the container
```console
$ cd /path/to/pgf
$ podman run -it --rm -v "$PWD:/__w:z" -w "/__w" registry.gitlab.com/islandoftex/images/texlive:latest l3build check -c config-examples -e luatex pgfmanual-en-library-decorations
```

I ran the tests and obtained this `diff` file:

<details>

<summary><code>./build/test-config-examples/pgfmanual-en-library-decorations.luatex.diff</code></summary>

```diff
*** ./build/test-config-examples/pgfmanual-en-library-decorations.tlg   2026-07-05 08:59:25.936735529 +0000
--- ./build/test-config-examples/pgfmanual-en-library-decorations.luatex.log    2026-07-05 08:59:56.222127212 +0000
***************
*** 5603,5611 ****
  ......\glue(\spaceskip) 0.0
  ......\pdfliteral origin{q }
  ......\hbox(0.0+0.0)x0.0, direction TLT
! .......\glue -4.50296
! .......\hbox(0.0+0.0)x0.0, shifted -0.28139, direction TLT
! ........\pdfliteral origin{0.0 1.0 -1.0 0.0 0.0 0.0 cm }
  ........\pdfliteral origin{q }
  ........\hbox(0.0+0.0)x0.0, direction TLT
  .........\hbox(3.47+0.97)x27.88998, direction TLT
--- 5603,5611 ----
  ......\glue(\spaceskip) 0.0
  ......\pdfliteral origin{q }
  ......\hbox(0.0+0.0)x0.0, direction TLT
! .......\glue 4.50299
! .......\hbox(0.0+0.0)x0.0, shifted -28.17137, direction TLT
! ........\pdfliteral origin{0.0 -1.0 1.0 0.0 0.0 0.0 cm }
  ........\pdfliteral origin{q }
  ........\hbox(0.0+0.0)x0.0, direction TLT
  .........\hbox(3.47+0.97)x27.88998, direction TLT
```

</details>

This is somewhere in the middle of `pgfmanual-en-library-decorations-19` which I found out by opening `./build/test-config-examples/pgfmanual-en-library-decorations.tlg` and scrolling upward from line 5603 to the beginning of the test.

Next up, I extract the single test from the large `pgfmanual-en-library-decorations.lvt` into a separate `test.tex` with some mocking of the `BOXTEST` framework and an additional `\box\PGFTESTBOX` to flush the result to the PDF for visual inspection.

<details>

<summary><code>test.tex</code></summary>

```latex
\documentclass{article}
\usepackage{tikz}
\usetikzlibrary{decorations,decorations.pathreplacing}
\begin{document}

\newbox\PGFTESTBOX

\message{^^JSTART-TEST-LOG^^J}

% inline the definition of \BEGINBOXTEST...\ENDBOXTEST
\setbox\PGFTESTBOX=\hbox\bgroup
\begin{tikzpicture}[>=stealth, every node/.style={midway, sloped, font=\tiny},
  decoration={show path construction,
    moveto code={
      \fill [red] (\tikzinputsegmentfirst) circle (2pt)
        node [fill=none, below] {moveto};},
    lineto code={
      \draw [blue,->] (\tikzinputsegmentfirst) -- (\tikzinputsegmentlast)
        node [above] {lineto};
    },
    curveto code={
      \draw [green!75!black,->] (\tikzinputsegmentfirst) .. controls
        (\tikzinputsegmentsupporta) and (\tikzinputsegmentsupportb)
        ..(\tikzinputsegmentlast) node [above] {curveto};
    },
    closepath code={
      \draw [orange,->] (\tikzinputsegmentfirst) -- (\tikzinputsegmentlast)
        node [above] {closepath};}
  }]
  \draw [help lines] grid (3,2);
  \path [decorate] (0,0) -- (3,1) arc (0:180:1.5 and 1) -- cycle;
\end{tikzpicture}
\egroup
\begingroup
  \showboxbreadth=\maxdimen
  \showboxdepth=\maxdimen
  \showbox\PGFTESTBOX
\endgroup

% actually flush the box for visual inspection
\box\PGFTESTBOX

\message{^^JEND-TEST-LOG^^J}

\end{document}
```

</details>

And finally here is the visual difference of the failing test.

| **Before** | **After** |
| --- | --- |
| before | after |

**Comment by Qrrbrbirlbel** [comment]
Thank you for the hand-holding.

> And finally here is the visual difference of the failing test.
>
> | **Before** | **After** |
> | --- | --- |
> | before | after |

Ah, a classic. The underlying problem is basically that for a vertical line (as the close-path segment) there's no true right-side up because the text is always absolutely orthogonal however you rotate it sloped.

Basically, which one is better?
| **Current** | **Flipped** |
| --- | --- |
| slopedtest | slopedtest |

I don't know if muzimuzhi's original fix changed this but this can easily be corrected to the previous output.

The vertical is indicated by `\pgf@x` being 0pt when this is happening ([line 474](https://github.com/Qrrbrbirlbel/pgf/blob/8ad294e445215685410cfcf96f0d9c0fe0bf531d/tex/generic/pgf/basiclayer/pgfcoretransformations.code.tex#L474-L477)):
```
\ifdim\pgf@x<0pt
  \pgf@x=-\pgf@x
  \pgf@y=-\pgf@y
\fi
```

I've flipped [muzimuzhi's check](https://github.com/pgf-tikz/pgf/issues/1058#issuecomment-940702317) for `\pgf@x < 0pt` into a `\pgf@x ≤ 0pt` check by doing `\unless\ifdim\pgf@x>0pt`:
```
\unless\ifdim\pgf@x>0pt
  \pgf@x=-\pgf@x
  \pgf@y=-\pgf@y
\fi
```

Even though PGF requires eTeX for some time now I haven't found one usage of `\unless` so we could also do which requires less mental logic by the reader.
```
\ifdim\pgf@x>0pt
\else
  \pgf@x=-\pgf@x
  \pgf@y=-\pgf@y
\fi
```

---

I've also fixed some typos in the manual and the comments where the `n` in <code>\ifpgfallowupsidedow<b>n</b>attime</code> was missing.

**Comment by Qrrbrbirlbel** [comment]
@hmenke 
I believe my recent change to this should be reverted.

With the **old** version
```
\documentclass{standalone}
\usepackage{tikz}
\begin{document}
\begin{tikzpicture}[midway, sloped, above]
  \draw[ultra thick]                      (0,1) -- node {AAA} (0,0);
  \draw[green]      (3,1) arc (0:180:1.5 and 1) -- node {BBB} (0,0);
\end{tikzpicture}
\end{document}
```
the result is
pgfsloped-deco

(This is the part of that decoration example that matters here.)

In fact, the dimens `\pgf@xc` and `\pgf@yc` before the necessary test is done in
https://github.com/pgf-tikz/pgf/blob/dc56af18b72255bede59800928f9a2bab1fe35e0/tex/generic/pgf/basiclayer/pgfcoretransformations.code.tex#L484-L487
will be

* 0.0pt and -28.45274p in the `AAA` version and
* -0.00002pt and -28.45274pt in the `BBB` version.

Clearly, the `arc` operation does not perfectly end at `(0, 1)`/`(0cm, 1cm)` as it ought to.
This actually leads `\pgftransformlineattime` to think that the end of the arc is slightly to the right of `(0, 0)`, thus `sloped` leads to `BBB` rotated that way.

---

muzimuzhi's original fix moved `\pgfpointnormalized` before this test and that turned the vector (`-0.00002pt`, `-28.45274pt`) into (`0.0pt`, `-1.0pt`), removing the imprecision of the `arc` operation but also leading to the weird `sloped` case which is perfectly vertical:

pgfsloped-deco

This might not look as nice in the decoration example but is in this case the expected behavior.

---

This all reminds me of various `0.05pt` values `\tikz@auto@anchor` tests for:
https://github.com/pgf-tikz/pgf/blob/dc56af18b72255bede59800928f9a2bab1fe35e0/tex/generic/pgf/frontendlayer/tikz/tikz.code.tex#L4486-L4510
which also tries to catch some close to horizontal/vertical angles.

---

So in summary:

1. arcs are imprecise.
2. the decoration example was accidentally easy to typeset.
3. muzimuzhi's original fix removed the imprecision making the result actually more correct.

I want to revert to the previous version. The question is then what we do about the decoration example:
1. Leave it as it is which results to a different diagram.
2. Changing the node placement so that it is to the left again (e.g. `rotate = 180`).
3. Changing the example so that there's no vertical closepath (e.g. `arc (0:180:0.75 and 1)` / `arc[start angle=0, end angle=180, x radius=0.75, y radius=1]`).

**Comment by hmenke** [comment]
I would like to avoid breaking the testcase (which includes modifying the testcase to do something else and make this pass). There are people out there who will have diagrams that will be broken by this change.

**Comment by Qrrbrbirlbel** [comment]
I've moved `\pgfpointnormalised` in `\pgftransform@common` down to after the mentioned check for negative `\pgf@x`.

I have now noticed that in the non-fixed version of `\pgftransformarcaxesattime`
https://github.com/pgf-tikz/pgf/blob/dc56af18b72255bede59800928f9a2bab1fe35e0/tex/generic/pgf/basiclayer/pgfcoretransformations.code.tex#L523-L535

the normalisation happens before the negativity check happens while the other two do it afterwards:
https://github.com/pgf-tikz/pgf/blob/dc56af18b72255bede59800928f9a2bab1fe35e0/tex/generic/pgf/basiclayer/pgfcoretransformations.code.tex#L479-L493

https://github.com/pgf-tikz/pgf/blob/dc56af18b72255bede59800928f9a2bab1fe35e0/tex/generic/pgf/basiclayer/pgfcoretransformations.code.tex#L572-L586

I can adjust the `\pgftransform@common` macro to take another two arguments where one is empty and the other one does the normalisation and swap those for `\pgftransformarcaxesattime` to really replicate the previous behavior.
